lib/internal/modules/esm/hooks.js

```diff
@@ -499,26 +501,38 @@ class HooksProxy {
   #isReady = false;
 
   constructor() {
-    const { InternalWorker } = require('internal/worker');
-    MessageChannel ??= require('internal/worker/io').MessageChannel;
-
+    const { InternalWorker, hooksPort } = require('internal/worker');
     const lock = new SharedArrayBuffer(SHARED_MEMORY_BYTE_LENGTH);
     this.#lock = new Int32Array(lock);
 
-    this.#worker = new InternalWorker(loaderWorkerId, {
-      stderr: false,
-      stdin: false,
-      stdout: false,
-      trackUnmanagedFds: false,
-      workerData: {
-        lock,
-      },
-    });
-    this.#worker.unref(); // ! Allows the process to eventually exit.
-    this.#worker.on('exit', process.exit);
+    if (isMainThread) {
+      // Main thread is the only one that creates the internal single hooks worker
+      this.#worker = new InternalWorker(loaderWorkerId, {
```

Comment:
I think if someone calls `module.register(hooks)` from a worker without calling it first on main thread the hooks worker is never created.
I don't think this is the intended use case for customization hooks but code and doc wise it's possible.

Maybe best to clearly disallow this or rework the hooks thread creation/termination in a way to not depend on main thread first.

Reply:
If noop means no customization hooks in use after calling `module.register` it's maybe not what user expects. Likely worth to clarify in docs.

Reply:
> At first glance it looks it should be a safe nop 

Perhaps it should error?

Reply:
I almost commented that it does error—I thought that was the original implementation when Lenon and I added it last year; but I checked the current and it doesn't. Haven't had a chance to check the original PR.

Reply:
as far as I interpret it, calling `module.register` on a  worker is not supported and should not do anything. If the expectation is to have an error, we should write a red testcase to use as spec here, so that we know we are all on the same page about what is expected. @Flarna , @GeoffreyBooth can you maybe write something with the exact scenario you had in mind?
@JakobJingleheimer does it error in the current version on `main`? Can you point me to the code that would do that in the original (before this PR) implementation? If it was like this, we should not lose that functionality and paired with the mentioned new testcase it should be like that in the future as well.

Reply:
I think a noop is unexpected behavior. Either the new hooks should get registered, and apply to all threads, or an error should be thrown. To do neither would feel like a bug to users, I’d think.

If that’s the current behavior on `main` then I guess we can handle this in a follow-up PR, though it should be very simple I’d think and could be attached to this if we want. Basically, create a new error like we already have for `ERR_LOADER_CHAIN_INCOMPLETE` (which we should probably rename to `ERR_HOOK_CHAIN_INCOMPLETE`, but in a different PR) and import it where we handle `register`, and if we’re not running on the main thread then throw the error. We could call it `ERR_MODULE_REGISTER_MUST_BE_CALLED_FROM_MAIN_THREAD` to be excessively clear.

Reply:
I think a follow-up PR for sure. This one handles a specific issue, and the one raised here, whilst importing and valid, is separate.

The code whence the error would be thrown lives in `esm/loader.js` (the exported `register`).

Reply:
implementing the error if `module.register` is definitely not complicated and I could take it over on this branch. But the difficulty is not in that imo. It is  that we would have to differentiate between two cases:
a) `module.register` is called explicitly on the user worker thread, somewhere in the thread script
b) implicit calls of `module.register` because they are called by a script that was passed on startup of the process via `--import` or `--experimental-loader` command line options.
That's why I would leave that to be solved as an own issue as it is not entirely related to the thread reduction and deal with it separately. We could add a line in the documentation of `module.register` that states the intended usage and that it is does something only when called on the main thread. That does cover all the needed scenarios afaict

Reply:
I don't know what was the behavior before. Actually I think it's don't care as this PR shows that hooks in context of workers was somewhat broken/incomplete before so nothing we want to keep untouched.

To my understanding the decision was to go for a single set of hooks for all workers. I think registration of this hooks should be done once for all before the main script runs (or as very first thing). Adding/changing hooks somewhat in the middle sound wrong and a foot gun.

Agree it's likely best to move this part into a followup.

Reply:
I pushed a fix for the unregistration of worker threads upon exit and I will give the error message a try. See if it is maybe easier to achieve than I first expected. If it is, I would leave that for a followup

Reply:
It sounds like we should definitely handle the “calls to `register` from a worker thread” issue in a follow-up, as it’s probably not simple and therefore it would be good to get it in an isolated commit. I think ideally it wouldn’t error, it would actually succeed: so pass a message back to the main thread “hey, use this file to add some hooks to the chains.” As in, I think we _could_ make that work, and it would be a more logical behavior to the user than an error message.

Reply:
I would gladly participate in that discussion on the follow-up and try to find a solution. I can imagine a certain level of flexibility that might be too much with that approach, could also be that I overthink it but let's get that discussion started and see where it goes.